File: 프로그래머스/3/92345. 사라지는 발판/사라지는 발판.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int dr[] = {-1, 1, 0, 0};
int dc[] = {0, 0, -1, 1};

int R, C;
vector<vector<int>> board;

int dfs(bool is_a_turn, pair<int, int> pa, pair<int, int> pb) {
    
    // 승부가 난 경우 더 움직일 수 없으므로 0 리턴
    if (board[pa.first][pa.second] == 0 || board[pb.first][pb.second] == 0) {
        return 0;
    }
    
    // 이후 턴에서 최선의 수를 저장하기위한 변수
    // 이후 재귀에서 진다는 결과가 나오면 최대 이동 횟수를, 이긴다는 결과가 나오면 최소 이동 횟수를 저장
    int best_steps = 0;
    
    int r = (is_a_turn) ? pa.first : pb.first;
    int c = (is_a_turn) ? pa.second : pb.second;
    for(int i=0; i<4; i++) {
        int nr = r + dr[i];
        int nc = c + dc[i];
        int steps; // 이후 게임을 진행할 때의 이동 횟수
        
        // 이동 가능할 경우 이동
        if (nr >=0 && nc >=0 && nr < R && nc < C && board[nr][nc]) {
            board[r][c] = 0;
            if (is_a_turn)
                steps = dfs(!is_a_turn, make_pair(nr, nc), pb) + 1;
            else
                steps = dfs(!is_a_turn, pa, make_pair(nr, nc)) + 1;
            board[r][c] = 1;
            
            // 지는 경우에 대한 최선의 수를 저장하다가, 이기는 경우를 발견한 경우 -> 이기는게 최선이므로 업데이트
            if (best_steps%2 == 0 && steps%2 == 1) {
                best_steps = steps;
            }
            // 지는 경우에 대한 최선의 수를 저장하다가, 지는 경우를 발견한 경우 -> 더 많이 움직인 쪽으로 업데이트
            else if (best_steps%2 == 0 && steps%2 == 0) {
                best_steps = max(best_steps, steps);
            }
            // 이기는 경우에 대한 최선의 수를 저장하다가, 이기는 경우를 발견한 경우 -> 더 적게 움직인 쪽으로 업데이트
            else if (best_steps%2 == 1 && steps%2 == 1) {
                best_steps = min(best_steps, steps);
            }
            
        }
    }
    
    // 위 for문에 걸리지 않았다면 사방으로 못 움직이는 것. best_steps가 0인 채로 리턴
    return best_steps;
}

int solution(vector<vector<int>> _board, vector<int> aloc, vector<int> bloc) {
    board = _board;
    R = board.size();
    C = board[0].size();
    // 항상 A가 먼저 시작
    // A와 B가 그냥 번갈아 자유롭게 움직이면서, 승부가 났을 때 최선으로 움직인 경우를 카운트
    // 이겼을 때는 가장 적게 움직이고, 졌을 때는 가장 많이 움직여야 하므로 이를 처리할 조건이 필요
    
    pair<int, int> pa = make_pair(aloc[0], aloc[1]);
    pair<int, int> pb = make_pair(bloc[0], bloc[1]);
    int result = dfs(true, pa, pb);
    
    return result;
}

File: 프로그래머스/3/92345. 사라지는 발판/사라지는 발판.java
import java.util.*;

class Solution {
    
    static int R, C;
    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};
    
    static int[][] board;
    
    // 지금까지 양쪽이 움직인 횟수를 리턴
    static int dfs(boolean is_a_turn, int ar, int ac, int br, int bc) {
        
        // 승부가 정해지면 더 이상 움직이지 못하고 0 리턴
        if (board[ar][ac] == 0 || board[br][bc] == 0) {
            return 0;
        }
        
        // 최선의 수까지 진행된 차례 수를 저장
        int best = 0;
        
        int r = (is_a_turn) ? ar : br;
        int c = (is_a_turn) ? ac : bc;
        
        for(int i=0; i<4; i++) {
            int turns = 0;
            
            int nr = r + dr[i];
            int nc = c + dc[i];
            if (nr >=0 && nc >=0 && nr < R && nc < C && board[nr][nc] == 1) {
                board[r][c] = 0;
                if (is_a_turn)
                    turns = dfs(!is_a_turn, nr, nc, br, bc) + 1; // 이동하고 턴을 넘긴다. 이동횟수가 1 추가된다.
                else
                    turns = dfs(!is_a_turn, ar, ac, nr, nc) + 1; // 이동하고 턴을 넘긴다. 이동횟수가 1 추가된다.
                
                board[r][c] = 1; // 백트래킹
            }
            
            // best가 짝수로 저장돼있다 -> 지금껏 찾은 최선의 수가 자신의 패배를 기준으로 저장돼있다
            if (best%2 == 0) {
                // turns의 홀짝 여부에 따라 현재 플레이어(is_a_turn 기준 A, B)가 마지막까지 가서 이겼는지, 졌는지 알 수 있다
                if (turns%2 == 1) best = turns; // 이긴 경우를 찾았다 -> 바로 업데이트
                else best = Math.max(best, turns); // 진 또 다른 경우를 찾았다 -> 더 많이 움직인 경우로 업데이트
            }
            // best가 홀수로 저장돼있다 -> 지금껏 찾은 최선의 수가 승리를 기준으로 저장돼있다.
            else {
                // 패배한 경우를 찾았다 -> 무시한다. 이긴 경우가 있는데 지는 경우로 업데이트 할 이유가 없다.
                // 이긴 또 다른 경우를 찾았다 -> 더 적게 움직인 경우로 업데이트
                if (turns%2 == 1) best = Math.min(best, turns);
            }
            
        }
        
        // 현재 위치에서 사방으로 움직지 못한 경우 best는 0에서 업데이트되지 않는다. 패배와 같이 0이 리턴된다.
        return best;
    }
    
    public int solution(int[][] _board, int[] aloc, int[] bloc) {
        R = _board.length;
        C = _board[0].length;
        board = _board;
        
        // 1. 게임이 끝날 때까지 번갈아가며 수를 둔다. 이 떄 각자 사방으로 이동해보며 dfs한다.
        // 2. 게임이 끝난 경우, 몇 번 차례가 지나고 끝났는지를 통해 현재 플레이어의 승패 여부를 판단할 수 있다.
        // 3. 이긴 경우 최대한 적게 움직인 수를, 진 경우 최대한 많이 움직인 수를 재귀적으로 리턴해서 저장한다.
        int result = dfs(true, aloc[0], aloc[1], bloc[0], bloc[1]);
        return result;
    }
}